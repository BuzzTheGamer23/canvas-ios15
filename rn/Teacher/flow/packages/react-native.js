// @flow
export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}
