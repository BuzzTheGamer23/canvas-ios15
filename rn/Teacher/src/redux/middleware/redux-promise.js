// @flow

import { isFSA } from 'flux-standard-action'
import type { Middleware, MiddlewareAPI, Dispatch } from 'redux'

export type Action = {
  payload: any,
  pending?: boolean,
  error?: boolean,
}

function isPromise (val: any): boolean {
  return val && typeof val.then === 'function'
}

/*
 * Simple middleware heavily based of off redux-promise for handling FSA
 * promises:
 *
 *   - When a FSA action is received with a payload that is a promise, or with
 *     a payload with a `promise` member, the action is tagged with a
 *     `pending: true` field.
 *   - When the attached promise resolves, a copy of the action is re-dispatched
 *     with its `payload` set to the promised value.
 *   - When the attached promise rejects, a copy of the action is re-dispatched
 *     with its `payload` set to the error, and the `error: true` tag.
 *
 */
let promiseMiddleware = <S, A: Action>({ dispatch }: MiddlewareAPI<S, A>): Middleware<S, A> => {
  return (next: Dispatch<A>) => (action: A) => {
    let promise = action.payload && action.payload.promise
    if (isFSA(action) && isPromise(promise) && !action.error) {
      promise.then(
        result => {
          let payload = { ...action.payload, result }
          delete payload.promise
          return dispatch({ ...action, payload })
        },
        error => {
          let payload = { ...action.payload, error }
          return dispatch({ ...action, payload, error: true })
        }
      )
      next({
        ...action,
        pending: true,
      })
    } else {
      next(action)
    }
  }
}

export default promiseMiddleware
