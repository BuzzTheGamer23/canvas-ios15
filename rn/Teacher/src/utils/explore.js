/* @flow */

function hasProp (object: any, key: string, value: any): boolean {
  if (object instanceof Object) {
    const props = object.props
    if (props && props instanceof Object) {
      for (var prop in props) {
        if (prop === key && props[prop] === value) {
          return true
        }
      }
    }
  }

  return false
}

function selectByProp (object: any, propKey: string, value: any): ?any {
  if (!(typeof object === 'object')) {
    return null
  }

  if (hasProp(object, propKey, value)) {
    return object
  }

  let result = null

  for (var key in object) {
    if (object[key] instanceof Array) {
      object[key].forEach((e) => {
        result = selectByProp(e, propKey, value)
        if (result) {
          return result
        }
      })
    }

    result = selectByProp(object[key], propKey, value)
    if (result) {
      return result
    }
  }

  return result
}

class ComponentExplorer {
  component: any

  constructor (component: any) {
    this.component = component
  }

  selectByProp (propKey: string, value: any): ?any {
    return selectByProp(this.component, propKey, value)
  }

  selectByID (id: string): ?any {
    return this.selectByProp('testID', id)
  }
}

export default function explore (component: any): ComponentExplorer {
  return new ComponentExplorer(component)
}
