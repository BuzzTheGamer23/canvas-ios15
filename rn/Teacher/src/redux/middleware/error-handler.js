//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import { Alert } from 'react-native'
import type { MiddlewareAPI } from 'redux'
import i18n from 'format-message'
import loginVerify from '../../common/login-verify'
export const ERROR_TITLE: string = i18n('Unexpected Error')
export const ERROR_MESSAGE: string = i18n('There was an unexpected error. Please try again.')

const showGlobalErrorAlertIfNecessary = (error: any) => {
  // should only continue if:
  // 1. error is not an api response error
  // 2. error is an api response error and the status code is 500 or above and the user is online
  let isOfflineError = error.message === 'Network Error'
  if (error.response != null) {
    error = error.response
  }
  if (!isOfflineError) {
    let errorMessage = parseErrorMessage(error)
    Alert.alert(ERROR_TITLE, errorMessage)
  }
}

const errorHandlerMiddleware: MiddlewareAPI = () => {
  return next => action => {
    if (action.error) {
      let error = action.payload.error
      if (error && error.response && error.response.status === 401) {
        loginVerify().then((invalid) => {
          if (!invalid) {
            if (action.payload.handlesError) return next(action)
            showGlobalErrorAlertIfNecessary(error)
          }
          next(action)
        })
        return // Bail out of this because of the promise. Next will be called after the promise has finished
      } else {
        if (action.payload.handlesError) return next(action)
        showGlobalErrorAlertIfNecessary(error)
      }
    }
    next(action)
  }
}

export default errorHandlerMiddleware

export function parseErrorMessage (error: any): string {
  if (error instanceof Error) {
    return error.message
  }

  if (error && error.data && error.data.errors && error.data.errors.length > 0) {
    return error.data.errors
        .map(error => error.message)
        .map(message => message.replace(/\.+$/, ''))
        .join('. ')
  } else if (error && error.data && error.data.errors instanceof Object && Object.keys(error.data.errors).length > 0) {
    let data = error.data.errors
    let result = Object.keys(data)
      .map((key, index) => data[key])
      .map(obj => obj.length > 0 ? obj[0] : {})
      .map(obj => `${obj.attribute || ''} ${obj.message || ''}`)
      .join('. ')

    if (result.trim()) {
      return result
    }
  }

  return ERROR_MESSAGE
}
