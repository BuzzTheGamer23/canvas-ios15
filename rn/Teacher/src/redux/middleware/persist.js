//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow
import { AsyncStorage } from 'react-native'
import { getSession } from 'canvas-api'
import type { MiddlewareAPI } from 'redux'
import hydrate from '../hydrate-action'

const STORE_VERSION = '3'

function storeName (session: Session): string {
  return `redux.state.${session.user.id}.${STORE_VERSION}`
}

async function removeOldStates (session: Session) {
  await AsyncStorage.multiRemove(
    (await AsyncStorage.getAllKeys())
      .filter(k => k.startsWith(`redux.state.${session.user.id}`))
  )
}

async function persistState (store) {
  let state = store.getState()
  let session = getSession()
  if (!session) return

  let expires = new Date()
  expires.setDate(expires.getDate() + 1)
  let cache = {
    expires: expires,
    state,
  }

  await AsyncStorage.setItem(storeName(session), JSON.stringify(cache))
}

export async function hydrateStoreFromPersistedState (store: any): any {
  let session = getSession()
  if (session) {
    let cachedState = await AsyncStorage.getItem(storeName(session))
    if (!cachedState) {
      await removeOldStates(session)
    }
    let appState
    try {
      appState = JSON.parse(cachedState)
    } catch (err) {}
    store.dispatch(hydrate(appState))
  }
}

const createPersistMiddleware = (timeoutWait: number): MiddlewareAPI => {
  return (store) => {
    let timeout
    return next => async (action) => {
      next(action)

      clearTimeout(timeout)
      timeout = setTimeout(() => persistState(store), timeoutWait)
    }
  }
}

export default createPersistMiddleware
