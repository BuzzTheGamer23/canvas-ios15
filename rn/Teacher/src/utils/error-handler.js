// @flow

import { Alert } from 'react-native'
import type { MiddlewareAPI } from 'redux'
import i18n from 'format-message'

export const ERROR_TITLE = i18n({
  default: 'Unexpected Error',
  description: 'The generic title of the generic error message',
})
export const ERROR_MESSAGE = i18n({
  default: 'There was an unexpected error. Please close this alert and try again.',
  description: 'A generic error message',
})

const errorHandlerMiddleware: MiddlewareAPI = () => {
  return next => action => {
    if (action.error) {
      // should only continue if:
      // 1. error is not an api response error
      // 2. error is an api response error and the status code is 500 or above
      // 3. the action hasn't indicated it will handle the error itself
      let error = action.payload.error || action.payload
      let isApiError = error.data && error.status
      if (!isApiError || error.status >= 500 || !action.payload.handlesError) {
        let errorMessage = error.message || ERROR_MESSAGE
        if (error.data && error.data.errors && error.data.errors.length > 0) {
          errorMessage = error.data.errors[0].message
        }
        Alert.alert(ERROR_TITLE, errorMessage)
      }
    }
    next(action)
  }
}

export default errorHandlerMiddleware
