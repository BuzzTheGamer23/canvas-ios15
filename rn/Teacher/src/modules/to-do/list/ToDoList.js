//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import { connect } from 'react-redux'
import {
  View,
  StyleSheet,
  FlatList,
  Alert,
  Image,
} from 'react-native'
import Screen from '../../../routing/Screen'
import color from '../../../common/colors'
import images from '../../../images'
import branding from '../../../common/branding'
import Actions from './actions'
import canvas from '../../../canvas-api'
import ToDoListItem from './ToDoListItem'
import RowSeparator from '../../../common/components/rows/RowSeparator'
import { ERROR_TITLE, parseErrorMessage } from '../../../redux/middleware/error-handler'
import { gradeProp } from '../../submissions/list/get-submissions-props'
import { Text } from '../../../common/text'
import i18n from 'format-message'

const { getToDo } = canvas

type OwnProps = {
  getToDo: () => Promise<ToDoItem[]>,
}

type StateProps = {
  items: ToDoItem[],
}

export type Props = OwnProps & StateProps & NavigationProps & typeof Actions

export class ToDoList extends Component<Props, any> {
  static defaultProps = {
    getToDo,
  }

  constructor (props: Props) {
    super(props)

    this.state = {
      refreshing: false,
      height: 0,
    }
  }

  // $FlowFixMe
  async componentWillMount () {
    await this.refresh()
  }

  render () {
    return (
      <Screen
        navBarColor={color.navBarColor}
        navBarButtonColor={color.navBarTextColor}
        navBarStyle='dark'
        drawUnderNavBar={true}
        navBarImage={branding.headerImage}
      >
        <View style={styles.container} onLayout={this.onLayout}>
          <FlatList
            data={this.props.items}
            renderItem={this.renderItem}
            refreshing={this.state.refreshing}
            onRefresh={() => { this.refresh() }}
            keyExtractor={(item, index) => String(index)}
            testID='to-do-list.list'
            ItemSeparatorComponent={RowSeparator}
            ListEmptyComponent={this.renderEmpty()}
          />
        </View>
      </Screen>
    )
  }

  renderItem = (({ item, index }: { item: ToDoItem, index: number }) => {
    return (
      <ToDoListItem
        item={item}
        index={index}
        onPress={this.onPressItem(item)}
      />
    )
  })

  refresh = async () => {
    this.setState({ refreshing: true })
    try {
      const { data } = await this.props.getToDo()
      this.props.refreshedToDo(data)
    } catch (error) {
      Alert.alert(ERROR_TITLE, parseErrorMessage(error))
    }
    this.setState({ refreshing: false })
  }

  onPressItem = (item: ToDoItem) => () => {
    this.showSpeedGrader(item)
  }

  showSpeedGrader = (item: ToDoItem) => {
    if (!item.course_id) return
    let assignmentID
    if (item.assignment) {
      assignmentID = item.assignment.id
    }
    if (item.quiz) {
      assignmentID = item.quiz.assignment_id
    }
    const path = `/courses/${item.course_id}/gradebook/speed_grader`
    const filter = (submissions) => submissions.filter(({ grade }) => grade === 'ungraded')
    this.props.navigator.show(
      path,
      { modal: true, modalPresentationStyle: 'fullscreen' },
      { filter, studentIndex: 0, assignmentID, onDismiss: this.refresh }
    )
  }

  onLayout = (event: any) => {
    this.setState({ height: event.nativeEvent.layout.height })
  }

  renderEmpty () {
    return (
      <View style={[styles.emptyContainer, { height: this.state.height }]}>
        <Image style={styles.emptyImage} source={images.relax} />
        <Text style={styles.emptyText}>{i18n('Nothing more to do.')}</Text>
        <Text style={styles.emptyText}>{i18n('Enjoy your day!')}</Text>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  emptyContainer: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'white',
  },
  emptyImage: {
    marginBottom: 24,
  },
  emptyText: {
    color: color.grey4,
    fontSize: 18,
  },
})

export function mapStateToProps ({ entities, toDo }: AppState): StateProps {
  const items = toDo
    .items
    .filter(i => i.type === 'grading')
    .map(item => {
      // We can't trust the item's `needs_grading_count` because the server
      // appears to cache this value for at least a few minutes so we
      // have to calculate it ourselves.
      let submissionRefs = []
      if (item.assignment &&
        entities.assignments &&
        entities.assignments[item.assignment.id] &&
        entities.assignments[item.assignment.id].submissions) {
        submissionRefs = entities.assignments[item.assignment.id].submissions.refs
      }
      if (item.quiz &&
        entities.quizzes &&
        entities.quizzes[item.quiz.id] &&
        entities.quizzes[item.quiz.id].submissions) {
        submissionRefs = entities.quizzes[item.quiz.id].submissions.refs
      }

      let submissions = []
      if (entities.submissions) {
        submissions = submissionRefs
          .map(r => entities.submissions[r])
          .filter(s => s)
      }

      if (!submissions.length) {
        return item
      }

      const needsGradingCount = submissions
        .filter(s => gradeProp(s.submission) === 'ungraded')
        .filter(s => {
          // filter out submissions that were very recently graded
          const limit = 3 * 60 * 1000 // 3 minutes
          return s.lastGradedAt == null ||
            Date.now() > new Date(s.lastGradedAt + limit).getTime()
        })
        .length

      return {
        ...item,
        needs_grading_count: needsGradingCount,
      }
    })
    .filter(item => item.needs_grading_count && item.needs_grading_count > 0)

  return {
    items,
  }
}

const Connected = connect(mapStateToProps, Actions)(ToDoList)
export default (Connected: *)
