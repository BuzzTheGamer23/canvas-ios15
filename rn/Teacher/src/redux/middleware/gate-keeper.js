//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import type { MiddlewareAPI } from 'redux'
import { Alert } from 'react-native'
import { HYDRATE_ACTION } from '../hydrate-action'
import logoutAction from '../logout-action'
import { isFSA } from 'flux-standard-action'
import i18n from 'format-message'
import { isOnline } from '../../utils/online-status'

let isLoggedIn = false
let isAlertOpen = false

export function updateAlertState (state: boolean): void {
  isAlertOpen = state
}

const gateKeeperMiddleware: MiddlewareAPI = () => {
  return next => action => {
    if (action.type === HYDRATE_ACTION) {
      isLoggedIn = true
    }

    if (action.type === logoutAction.type) {
      isLoggedIn = false
    }

    let promise = action.payload && action.payload.promise || action.payload
    if (!isOnline() && isFSA(action) && promise instanceof Promise) {
      if (!isAlertOpen) {
        updateAlertState(true)
        Alert.alert(
          i18n('No internet connection'),
          i18n('This action requires an internet connection.'),
          [
            { text: 'Cancel', onPress: () => updateAlertState(false), style: 'cancel' },
          ]
        )
      }
    }

    if (isLoggedIn || action.type === logoutAction.type) {
      next(action)
    }
  }
}

export default gateKeeperMiddleware
