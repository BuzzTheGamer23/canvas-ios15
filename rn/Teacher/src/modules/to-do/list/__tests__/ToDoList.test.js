//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

/* @flow */

import React from 'react'
import { Alert } from 'react-native'
import renderer from 'react-test-renderer'
import { ToDoList, mapStateToProps, type Props } from '../ToDoList'
import { ERROR_TITLE } from '../../../../redux/middleware/error-handler'
import explore from '../../../../../test/helpers/explore'

jest
  .mock('Button', () => 'Button')
  .mock('TouchableHighlight', () => 'TouchableHighlight')
  .mock('TouchableOpacity', () => 'TouchableOpacity')
  .mock('../ToDoListItem', () => 'ToDoListItem')

const template = {
  ...require('../../../../__templates__/toDo'),
  ...require('../../../../__templates__/helm'),
  ...require('../../../../__templates__/error'),
  ...require('../../../../__templates__/assignments'),
  ...require('../../../../__templates__/quiz'),
  ...require('../../../../__templates__/submissions'),
  ...require('../../../../redux/__templates__/app-state'),
}

describe('ToDoList', () => {
  let props: Props
  beforeEach(() => {
    props = {
      navigator: template.navigator(),
      items: [template.toDoItem()],
      getToDo: jest.fn(() => Promise.resolve({ data: [template.toDoItem()] })),
      refreshedToDo: jest.fn(),
    }
  })

  it('renders items', () => {
    props.items = [template.toDoItem()]
    expect(render(props).toJSON()).toMatchSnapshot()
  })

  it('refreshes on mount', async () => {
    const spy = jest.fn(() => Promise.resolve({ data: [template.toDoItem()] }))
    props.getToDo = spy
    const view = render(props)
    await view.getInstance().componentWillMount()
    expect(spy).toHaveBeenCalled()
  })

  it('dispatches refreshedToDo on refresh', async () => {
    const spy = jest.fn()
    props.refreshedToDo = spy
    const data = [template.toDoItem()]
    props.getToDo = jest.fn(() => Promise.resolve({ data }))
    const view = render(props)
    const list: any = explore(view.toJSON()).selectByType('RCTScrollView')
    await list.props.onRefresh()
    expect(spy).toHaveBeenCalledWith(data)
  })

  it('alerts refresh error', async () => {
    const spy = jest.fn()
    // $FlowFixMe
    Alert.alert = spy
    props.getToDo = jest.fn(() => Promise.reject(template.error('ERROR')))
    const view = render(props)
    const list: any = explore(view.toJSON()).selectByType('RCTScrollView')
    await list.props.onRefresh()
    expect(spy).toHaveBeenCalledWith(ERROR_TITLE, 'ERROR')
  })

  it('shows Speed Grader on assignment press', () => {
    const spy = jest.fn()
    props.navigator = template.navigator({ show: spy })
    props.items = [template.toDoItem({
      course_id: '222',
      assignment: template.assignment({ id: '1' }),
    })]
    const view = render(props)
    const row: any = explore(view.toJSON()).selectByType('ToDoListItem')
    row.props.onPress()
    expect(spy).toHaveBeenCalledWith(
      '/courses/222/gradebook/speed_grader',
      { modal: true, modalPresentationStyle: 'fullscreen' },
      {
        filter: expect.any(Function),
        studentIndex: 0,
        assignmentID: '1',
        onDismiss: expect.any(Function),
      },
    )
  })

  it('shows Speed Grader on quiz press', () => {
    const spy = jest.fn()
    props.navigator = template.navigator({ show: spy })
    props.items = [template.toDoItem({
      course_id: '222',
      quiz: template.quiz({ id: '2', assignment_id: '1' }),
    })]
    const view = render(props)
    const row: any = explore(view.toJSON()).selectByType('ToDoListItem')
    row.props.onPress()
    expect(spy).toHaveBeenCalledWith(
      '/courses/222/gradebook/speed_grader',
      { modal: true, modalPresentationStyle: 'fullscreen' },
      {
        filter: expect.any(Function),
        studentIndex: 0,
        assignmentID: '1',
        onDismiss: expect.any(Function),
      },
    )
  })

  function render (props: Props, options: any = {}): any {
    return renderer.create(<ToDoList {...props} />, options)
  }
})

describe('mapStateToProps', () => {
  it('uses counts directly from toDo entities if there are no submissions', () => {
    const items = [template.toDoItem({ assignment: template.assignment() })]
    const state = template.appState({
      toDo: { items },
      entities: {
        assignments: {},
      },
    })
    expect(mapStateToProps(state)).toEqual({ items })
  })

  it('uses submissions to determine needs_grading_count', () => {
    const assignment = template.assignment({ id: '1' })
    const ungraded = template.submission({
      id: '1',
      grade: null,
    })
    const graded = template.submission({
      id: '2',
      grade: 'A',
    })
    const staleItem = template.toDoItem({
      assignment,
      needs_grading_count: 2,
    })
    const state = template.appState({
      toDo: { items: [staleItem] },
      entities: {
        assignments: {
          '1': {
            data: assignment,
            submissions: {
              pending: 0,
              refs: [ungraded.id, graded.id],
            },
          },
        },
        submissions: {
          [ungraded.id]: {
            submission: ungraded,
          },
          [graded.id]: {
            submission: graded,
          },
        },
      },
    })
    expect(mapStateToProps(state)).toEqual({
      items: [{ ...staleItem, needs_grading_count: 1 }],
    })
  })

  it('it does not include recently graded submissions', () => {
    const quiz = template.quiz({ id: '1' })
    const s1 = template.submission({
      id: '1',
      grade: null,
    })
    const s2 = template.submission({
      id: '2',
      grade: null,
    })
    const staleItem = template.toDoItem({
      quiz,
      needs_grading_count: 2,
    })
    const state = template.appState({
      toDo: { items: [staleItem] },
      entities: {
        quizzes: {
          '1': {
            data: quiz,
            submissions: {
              pending: 0,
              refs: [s1.id, s2.id],
            },
          },
        },
        submissions: {
          [s1.id]: {
            submission: s1,
          },
          [s2.id]: {
            submission: s2,
            lastGradedAt: new Date(Date.now() - 60000).getTime(), // 1 minute ago
          },
        },
      },
    })
    expect(mapStateToProps(state)).toEqual({
      items: [{ ...staleItem, needs_grading_count: 1 }],
    })
  })
})
