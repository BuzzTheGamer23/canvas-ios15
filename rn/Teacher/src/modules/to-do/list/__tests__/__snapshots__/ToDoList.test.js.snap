// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList renders items 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
            },
            Object {
              "height": 0,
            },
          ]
        }
      >
        <Image
          source={1}
          style={
            Object {
              "marginBottom": 24,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#8B969E",
              "fontSize": 18,
            }
          }
        >
          Nothing more to do.
        </Text>
        <Text
          style={
            Object {
              "color": "#8B969E",
              "fontSize": 18,
            }
          }
        >
          Enjoy your day!
        </Text>
      </View>
    }
    data={
      Array [
        Object {
          "assignment": Object {
            "assignment_group_id": 1376,
            "automatic_peer_reviews": false,
            "course_id": "987654321",
            "created_at": "2017-03-17T19:15:25Z",
            "description": "a course that can take any beginner to professional in a matter of microseconds.  Harness the power of your favorite languages like PHP and Cold Fusion.",
            "due_at": "2037-06-01T05:59:00Z",
            "grading_type": "points",
            "html_url": "http://hostname.tld/whatever",
            "id": "123456789",
            "lock_at": "2037-06-01T05:59:59Z",
            "name": "learn to write code",
            "needs_grading_count": 0,
            "overrides": Array [],
            "peer_reviews": false,
            "points_possible": 20,
            "position": 1,
            "submission_types": Array [
              "online_text_entry",
              "online_url",
              "media_recording",
              "online_upload",
            ],
            "unlock_at": "2017-03-17T06:00:00Z",
            "updated_at": "2017-03-17T19:15:25Z",
          },
          "context_type": "course",
          "course_id": "1",
          "group_id": null,
          "html_url": "https://api.instructure.com/api/v1/path/to/html_url",
          "ignore": "https://api.instructure.com/api/v1/path/to/ignore",
          "ignore_permanently": "https://api.instructure.com/api/v1/path/to/ignore_permanently",
          "needs_grading_count": 5,
          "type": "grading",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={undefined}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        progressViewOffset={undefined}
        refreshing={true}
      />
    }
    refreshing={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="to-do-list.list"
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <ToDoListItem
          index={0}
          item={
            Object {
              "assignment": Object {
                "assignment_group_id": 1376,
                "automatic_peer_reviews": false,
                "course_id": "987654321",
                "created_at": "2017-03-17T19:15:25Z",
                "description": "a course that can take any beginner to professional in a matter of microseconds.  Harness the power of your favorite languages like PHP and Cold Fusion.",
                "due_at": "2037-06-01T05:59:00Z",
                "grading_type": "points",
                "html_url": "http://hostname.tld/whatever",
                "id": "123456789",
                "lock_at": "2037-06-01T05:59:59Z",
                "name": "learn to write code",
                "needs_grading_count": 0,
                "overrides": Array [],
                "peer_reviews": false,
                "points_possible": 20,
                "position": 1,
                "submission_types": Array [
                  "online_text_entry",
                  "online_url",
                  "media_recording",
                  "online_upload",
                ],
                "unlock_at": "2017-03-17T06:00:00Z",
                "updated_at": "2017-03-17T19:15:25Z",
              },
              "context_type": "course",
              "course_id": "1",
              "group_id": null,
              "html_url": "https://api.instructure.com/api/v1/path/to/html_url",
              "ignore": "https://api.instructure.com/api/v1/path/to/ignore",
              "ignore_permanently": "https://api.instructure.com/api/v1/path/to/ignore_permanently",
              "needs_grading_count": 5,
              "type": "grading",
            }
          }
          onPress={[Function]}
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
